### PR TITLE
simplify the ImageDecoder native code binding

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

File: sky/engine/core/dart/painting.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// found in the LICENSE file.
+
+part of dart_ui;
+
+typedef void _ImageDecoderCallback(Image result);
+
+class PaintingNatives {
+  static void decodeFromDataPipe(int handle, _ImageDecoderCallback callback)
+      native "PaintingNatives_decodeFromDataPipe";
+}
```
Comment (abarth) [inline sky/engine/core/dart/painting.dart]: There no need for this PaintingNatives class.  Instead, we can just have a top-level function called `decodeImageFromDataPipe`

File: sky/engine/core/painting/painting.cc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include "sky/engine/platform/mojo/data_pipe.h"
+#include "sky/engine/tonic/dart_invoke.h"
+#include "sky/engine/tonic/dart_persistent_value.h"
+#include "sky/engine/tonic/mojo_converter.h"
+
+namespace blink {
+namespace {
+
+void DecodeImage(scoped_ptr<DartPersistentValue> callback,
+		 PassRefPtr<SharedBuffer> buffer) {
```
Comment (abarth) [inline sky/engine/core/painting/painting.cc]: Missing one space in this indent.  :)

File: sky/engine/core/painting/painting.cc [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  DartState::Scope scope(dart_state);
+
+  OwnPtr<ImageDecoder> decoder =
+      ImageDecoder::create(*buffer.get(), ImageSource::AlphaPremultiplied,
+                           ImageSource::GammaAndColorProfileIgnored);
+
+  // decoder can be null if the buffer was empty and we couldn't even guess
+  // what type of image to decode.
+  if (!decoder) {
+    DartInvokeAppClosure(callback->value(), {nullptr});
```
Comment (abarth) [inline sky/engine/core/painting/painting.cc]: s/nullptr/Dart_Null()/

File: sky/engine/core/painting/painting.cc [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    return;
+  }
+  decoder->setData(buffer.get(), true);
+  if (decoder->failed() || decoder->frameCount() == 0) {
+    DartInvokeAppClosure(callback->value(), {nullptr});
```
Comment (abarth) [inline sky/engine/core/painting/painting.cc]: ditto

File: sky/engine/core/painting/painting.cc [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  RefPtr<CanvasImage> resultImage = CanvasImage::create();
+  ImageFrame* imageFrame = decoder->frameBufferAtIndex(0);
+  RefPtr<SkImage> skImage = adoptRef(
+      SkImage::NewFromBitmap(imageFrame->getSkBitmap()));
+  resultImage->setImage(skImage.release());
+
+  Dart_Handle resultHandle = DartConverter<CanvasImage*>::ToDart(
+      resultImage.get());
```
Comment (abarth) [inline sky/engine/core/painting/painting.cc]: You can just write `ToDart(resultImage.get())`.  There's template magic to automatically detect the type.

File: sky/engine/core/painting/painting.cc [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  DartInvokeAppClosure(callback->value(), {resultHandle});
```
Comment (abarth) [inline sky/engine/core/painting/painting.cc]: I would just inline the `ToDart` call here and skip the local variable.

**Comment by abarth** [comment]
LGTM

**Comment by jason-simmons** [comment]
Addressed feedback and added the ui.decodeImageFromList API

**Comment by abarth** [comment]
LGTM

**Pull request merged**